module_2/gifts/service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+package com.epam.esm.service.impl;
+
+import java.time.LocalDateTime;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.MultiValueMap;
+
+import com.epam.esm.dto.CertificateDto;
+import com.epam.esm.dto.TagDto;
+import com.epam.esm.exception.ErrorCode;
+import com.epam.esm.exception.NotFoundException;
+import com.epam.esm.repository.CertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.repository.model.CertificateModel;
+import com.epam.esm.repository.model.TagModel;
+import com.epam.esm.repository.query_builder.EntityConstant;
+import com.epam.esm.service.CertificateService;
+import com.epam.esm.service.TagService;
+import com.epam.esm.service.converter.CertificateConverter;
+import com.epam.esm.service.converter.TagConverter;
+import com.epam.esm.service.validation.CertificateValidation;
+import com.epam.esm.service.validation.TagValidation;
+import com.epam.esm.service.validation.Util;
+
+/**
+ * 
+ * Contains methods implementation for working mostly with {@code CertificateDto}
+ * entity.
+ *
+ */
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+	@Autowired
+	private CertificateRepository certificateRepository;
+
+	@Autowired
+	private TagRepository tagRepository;
```

Comment:
Why do you use in GiftCertificateService TagRepository? It's better to call tagService methods